fix: datepicker doesn't close on first selection of date on month change

SaiCharanChetpelly31 opened the pull request:
Fixes issue #30701 

Added cypress test for the fix

https://github.com/appsmithorg/appsmith/assets/119920490/d8e92377-6d27-49f9-8839-e18b8b6aae65



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Added new tests to the DatePicker widget ensuring functionality and reliability.
- **Bug Fixes**
	- Improved date selection logic to handle null values and formatted dates correctly.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SaiCharanChetpelly31 commented:
Hi @ApekshaBhosale Can you please review this PR?

SaiCharanChetpelly31 commented:
@ApekshaBhosale   Gentle Reminder: Could you please review this PR?

SaiCharanChetpelly31 commented:
Hi @ApekshaBhosale  ,

I hope you're well. I’m curious and excited to hear your thoughts on my pull request raised a while ago. Looking forward to your review!
Thank you.

ApekshaBhosale commented on app/client/src/widgets/DatePickerWidget2/component/index.tsx:
```diff
@@ -481,6 +481,18 @@ class DatePickerComponent extends React.Component<
     }
   };
+
+  handleDateChange = (selectedDate: Date | null) => {
+    if (selectedDate === null) {
+      this.setState({ selectedDate: "" });
+      this.props.onDateSelected("");
+    } else {
+      const formattedDate = moment(selectedDate).format(ISO_DATE_FORMAT);
+      this.setState({ selectedDate: formattedDate }, () => {
+        this.props.onDateSelected(formattedDate);
+      });
+    }
+  };
```
@SaiCharanChetpelly31 please consider this to make change to  your code

ApekshaBhosale commented on app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    it("1.should close datepicker on single click when month is changed", function () {
+      cy.openPropertyPane("datepickerwidget2");
```
use locators from WigetLocator file

ApekshaBhosale commented on app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:
```diff
@@ -0,0 +1,43 @@
+    it("1.should close datepicker on single click when month is changed", function () {
+      cy.openPropertyPane("datepickerwidget2");
+      cy.get(".t--widget-datepickerwidget2 input").click();
+      cy.get(".DayPicker").should("be.visible");
```
use `AssertElementVisibility` helper

ApekshaBhosale commented on app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:
```diff
@@ -0,0 +1,43 @@
+      cy.get(".DayPicker").should("be.visible");
+
+      navigateAndSelectRandomDate("prev");
+      cy.get(".t--widget-datepickerwidget2 input").click();
```
use `GetNClick`  helper

ApekshaBhosale commented on app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:
```diff
@@ -0,0 +1,43 @@
+      navigateAndSelectRandomDate("prev");
+      cy.get(".t--widget-datepickerwidget2 input").click();
+      navigateAndSelectRandomDate("prev");
+      cy.get(".t--widget-datepickerwidget2 input").click();
```
use `GetNClick` helper

SaiCharanChetpelly31 commented:
@ApekshaBhosale I accidentally deleted my forked repository, but I have raised another PR with the suggested changes addressing the comments. Could you please review [this PR](https://github.com/appsmithorg/appsmith/pull/34679)?
